src/github.com/getlantern/flashlight/client/chained.go

```diff
@@ -105,6 +128,117 @@ func (s *ChainedServerInfo) Dialer(deviceID string) (*balancer.Dialer, error) {
 
 			return conn, nil
 		},
+		Check: func() bool {
+			return s.check(d, deviceID)
+		},
 		OnRequest: ccfg.OnRequest,
 	}, nil
 }
+
+func (s *chainedServer) attachHeaders(req *http.Request, deviceID string) {
+	authToken := s.AuthToken
+	if ForceAuthToken != "" {
+		authToken = ForceAuthToken
+	}
+	if authToken != "" {
+		req.Header.Add("X-LANTERN-AUTH-TOKEN", authToken)
+	}
+	req.Header.Set("X-LANTERN-DEVICE-ID", deviceID)
+}
+
+func (s *chainedServer) check(dial func(string, string) (net.Conn, error), deviceID string) bool {
+	rt := &http.Transport{
+		DisableKeepAlives: true,
+		Dial:              dial,
+	}
+	var url string
+	checkTarget := s.checkTargets.get()
+	if checkTarget == "" {
+		url = "http://ping-chained-server"
+	} else {
+		url = fmt.Sprintf("http://%s/index.html", checkTarget)
+	}
+	req, err := http.NewRequest("HEAD", url, nil)
+	if err != nil {
+		log.Errorf("Could not create HTTP request: %v", err)
+		return false
+	}
+	if checkTarget == "" {
+		req.Header.Set("X-Lantern-Ping", "small")
+	}
+
+	s.attachHeaders(req, deviceID)
+	ok, timedOut, _ := withtimeout.Do(60*time.Second, func() (interface{}, error) {
+		resp, err := rt.RoundTrip(req)
+		if err != nil {
+			log.Debugf("Error testing dialer %s to %s: %s", s.Addr, url, err)
+			return false, nil
+		}
+		if err := resp.Body.Close(); err != nil {
+			log.Debugf("Unable to close response body: %v", err)
+		}
+		msg := fmt.Sprintf("HEAD %s through chained server at %s, status code %d", url, s.Addr, resp.StatusCode)
+		// < 500 means the check target is at least reachable through this
+		// chained server, no matter what the HTTP status code is.
+		//
+		// The only exception is that if chained server rejects current client
+		// because of invalid token, etc., we can't differentiate it from an
+		// status code from target site.
+		if reachable := resp.StatusCode < 500; !reachable {
+			log.Debug(msg)
+			return false, nil
+		}
+		log.Trace(msg)
+		if checkTarget != "" {
+			// can be used as check target again if no new sites is added
+			s.checkTargets.add(checkTarget)
+		}
+		return true, nil
+	})
+	if timedOut {
+		log.Errorf("Timed out checking dialer at: %v", s.Addr)
+	}
+	return !timedOut && ok.(bool)
+}
+
+func (s *chainedServer) addCheckTarget(addr string) {
+	host, port, e := net.SplitHostPort(addr)
+	if e != nil {
+		log.Errorf("failed to split port from %s", addr)
+		return
+	}
+	if port != "80" {
+		log.Tracef("Skip setting non-HTTP site %s as check target", addr)
```

Comment:
Haha, just saw this.  That addresses my previous comment about non-HTTP check targets :)